File: Greedy_Algorithms/JavaScript/huffman_coding.js
class Node {
    constructor(char, freq, left = null, right = null) {
        this.char = char;
        this.freq = freq;
        this.left = left;
        this.right = right;
    }
}

// Priority Queue implementation (simple array-based for demonstration)
class PriorityQueue {
    constructor() {
        this.collection = [];
    }

    enqueue(element) {
        this.collection.push(element);
        this.sort();
    }

    dequeue() {
        return this.collection.shift();
    }

    sort() {
        this.collection.sort((a, b) => a.freq - b.freq);
    }

    size() {
        return this.collection.length;
    }
}

function buildHuffmanTree(frequencies) {
    const pq = new PriorityQueue();

    for (const char in frequencies) {
        pq.enqueue(new Node(char, frequencies[char]));
    }

    while (pq.size() > 1) {
        const left = pq.dequeue();
        const right = pq.dequeue();

        const merged = new Node(null, left.freq + right.freq, left, right);
        pq.enqueue(merged);
    }

    return pq.dequeue();
}

function generateHuffmanCodes(node, currentCode, huffmanCodes) {
    if (node === null) {
        return;
    }

    if (node.char !== null) {
        huffmanCodes[node.char] = currentCode;
        return;
    }

    generateHuffmanCodes(node.left, currentCode + "0", huffmanCodes);
    generateHuffmanCodes(node.right, currentCode + "1", huffmanCodes);
}

function encode(text, huffmanCodes) {
    let encodedText = "";
    for (const char of text) {
        encodedText += huffmanCodes[char];
    }
    return encodedText;
}

function decode(encodedText, huffmanTree) {
    let decodedText = "";
    let current = huffmanTree;

    for (const bit of encodedText) {
        if (bit === '0') {
            current = current.left;
        } else {
            current = current.right;
        }

        if (current.char !== null) {
            decodedText += current.char;
            current = huffmanTree; // Reset to root for next character
        }
    }
    return decodedText;
}

function huffmanCoding(text) {
    // 1. Tính tần số của các ký tự
    const frequencies = {};
    for (const char of text) {
        frequencies[char] = (frequencies[char] || 0) + 1;
    }

    // 2. Xây dựng cây Huffman
    const huffmanTree = buildHuffmanTree(frequencies);

    // 3. Tạo mã Huffman
    const huffmanCodes = {};
    generateHuffmanCodes(huffmanTree, "", huffmanCodes);

    // 4. Mã hóa văn bản
    const encodedText = encode(text, huffmanCodes);

    return { encodedText, huffmanCodes, huffmanTree };
}

// Ví dụ sử dụng:
const text = "this is an example for huffman encoding";
const { encodedText, huffmanCodes, huffmanTree } = huffmanCoding(text);

console.log("Mã Huffman:", huffmanCodes);
console.log("Văn bản được mã hóa:", encodedText);

const decodedText = decode(encodedText, huffmanTree);
console.log("Văn bản được giải mã:", decodedText);
console.log("Văn bản gốc == Văn bản được giải mã:", text === decodedText);

File: Greedy_Algorithms/Python/huffman_coding.py

import heapq

class Node:
    def __init__(self, char, freq):
        self.char = char
        self.freq = freq
        self.left = None
        self.right = None

    # So sánh các đối tượng Node dựa trên tần số của chúng
    def __lt__(self, other):
        return self.freq < other.freq

def build_huffman_tree(frequencies):
    priority_queue = []
    for char, freq in frequencies.items():
        heapq.heappush(priority_queue, Node(char, freq))

    while len(priority_queue) > 1:
        left = heapq.heappop(priority_queue)
        right = heapq.heappop(priority_queue)

        merged = Node(None, left.freq + right.freq)
        merged.left = left
        merged.right = right

        heapq.heappush(priority_queue, merged)

    return priority_queue[0]

def generate_huffman_codes(node, current_code, huffman_codes):
    if node is None:
        return

    if node.char is not None:
        huffman_codes[node.char] = current_code
        return

    generate_huffman_codes(node.left, current_code + "0", huffman_codes)
    generate_huffman_codes(node.right, current_code + "1", huffman_codes)

def huffman_coding(text):
    # 1. Tính tần số của các ký tự
    frequencies = defaultdict(int)
    for char in text:
        frequencies[char] += 1

    # 2. Xây dựng cây Huffman
    huffman_tree = build_huffman_tree(frequencies)

    # 3. Tạo mã Huffman
    huffman_codes = {}
    generate_huffman_codes(huffman_tree, "", huffman_codes)

    # 4. Mã hóa văn bản
    encoded_text = "".join(huffman_codes[char] for char in text)

    return encoded_text, huffman_codes
